src/main/cli/process.ts

```diff
@@ -0,0 +1,68 @@
+// ==LICENSE-BEGIN==
+// Copyright 2017 European Digital Reading Lab. All rights reserved.
+// Licensed to the Readium Foundation under one or more contributor license agreements.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file exposed on Github (readium) in the project repository.
+// ==LICENSE-END==
+
+import * as debug_ from "debug";
+import { app } from "electron";
+import { ICli, ICliParam } from "readium-desktop/main/cli/commandLine";
+import { Arguments } from "yargs";
+
+declare const __APP_NAME__: string;
+declare const __APP_VERSION__: string;
+
+// Logger
+const debug = debug_("readium-desktop:cli");
+
+export async function processCommandLine(commandLine: ICli[], argv: Arguments): Promise<boolean> {
+
+    const param: ICliParam = {
+        argv,
+        quit: false,
+    };
+    debug("param: ", param);
+    let returnCode = 0;
+
+    // assign arg and remove $0
+    const arg = Object.keys(argv);
+    arg.splice(arg.indexOf("$0"), 1);
+
+    // print help and quit
+    if (arg.indexOf("help") > -1 || arg.indexOf("h") > -1) {
+        process.stdout.write(`${__APP_NAME__} ${__APP_VERSION__}\n`);
+        process.stdout.write("\n");
+        process.stdout.write(`Usage: ${argv.$0} [options][paths...]\n`);
+        process.stdout.write("\n");
+        process.stdout.write(`Options\n`);
+        commandLine.map((v) => {
+            if (v.help.length) {
+                process.stdout.write(` ${v.help[0]}\t\t${v.help[1] ? v.help[1] : ""}\n`);
+            }
+        });
+        param.quit = true;
+    } else {
+        // execute all commands
+        const commandPromiseTab = arg.map((command) => {
+            const op = commandLine.find((c) => c.name === command);
+            if (op && (op.name !== "_" || argv._.length)) {
+                    return op.fct(param);
+            }
+        });
+
+        try {
+            const commandRes = await Promise.all(commandPromiseTab);
+            const ifFalse = commandRes.indexOf(false);
+            returnCode = ifFalse < 0 ? 0 : 1;
+        } catch (e) {
+            returnCode = 1;
+        }
+    }
+
+    if (param.quit) {
+        app.exit(returnCode);
```

Comment:
This is fine in `--silent` mode, because no reader windows are opened.
Otherwise, we need to use `app.quit()` instead ( https://electronjs.org/docs/api/app#appquit ) so that the app subsystems are shutdown gracefully.
For example, there is a `app.on("will-quit", ...)` event listener in the "navigator" which handles webview sessions:
https://github.com/readium/r2-navigator-js/blob/95f15b54977b309ff3e14e971980753089fab0dc/src/electron/main/sessions.ts#L265

Reply:
@panaC said:

> we can use app.exit because no window is start

https://github.com/readium/readium-desktop/pull/403#issuecomment-512746474

Reply:
So `param.quit` is guaranteed to be true **only** when `--silent` is set. That's good.

Now, let's see what happens with:

1) `EXE --silent --read=TITLE` (in my case: `npm run cli -- --silent --read=Accessible` to open the existing "Accessible EPUB 3" book in the library)
2) `EXE PATH_TO_EPUB --silent` (in my case: `npm run cli -- /Volumes/EPUBs/childrens-literature.epub --silent`)

I tweaked your code so that the app is forcefully killed with `app.exit()` after a short timeout (plus a console message so I can see what happens):
```
    if (param.quit) {
        process.stderr.write(`KILL...\n`);
        setTimeout(() => {
            app.exit(returnCode);
        }, 3000);
        return true;
    }
```

In both cases 1) and 2), the app opens the publication successfully in the reader UI, and then the app is killed forcefully with `app.exit()` (no `app.quit()` clean destroy). In my test this happens after my 3s timeout, but in the current PR this occurs immediately (which is not acceptable either).

I suggest ignoring the `--silent` command line parameter when the app successfully opens its UI. Otherwise (e.g. missing TITLE, or missing PATH_TO_EPUB), then `app.exit()` with the appropriate failure code.